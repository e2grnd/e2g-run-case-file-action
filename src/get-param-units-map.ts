/* eslint-disable github/array-foreach */
import {TParams} from './load-calc-params'
import {TDataTables} from './load-calc-tables'
import {TUnitSystem} from './main'

export type UnitMap = Record<string, string | string[]>

export function getParamUnitsMap(params: TParams, dataTables: TDataTables, unitSystem: TUnitSystem): UnitMap {
  const dtUnits = dataTables.reduce((acc, t) => {
    if (t.keywords) {
      t.keywords.forEach(kw => {
        if (kw.units) {
          acc[kw.keyword] = kw.units[unitSystem]
        }
      })
    }
    return acc
  }, {} as UnitMap)
  return params.reduce((acc: UnitMap, p) => {
    if (p.units) {
      acc[p.keyword] = p.units[unitSystem]
    }
    return acc
  }, dtUnits)
}
