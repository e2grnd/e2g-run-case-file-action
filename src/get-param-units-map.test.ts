import {getParamUnitsMap, UnitMap} from './get-param-units-map'
import paramsBPCRaw from './fixtures/bpc/params'
import dataTablesBPC from './fixtures/bpc/dataTables'
import paramsCharpyRaw from './fixtures/charpy/params'
import dataTablesCharpy from './fixtures/charpy/dataTables'

const paramsBPC = paramsBPCRaw as any
const paramsCharpy = paramsCharpyRaw as any

describe('getParamUnitsMap', () => {
  it('returns the unit map correctly', () => {
    const unitSystem = 'Metric'

    const result = getParamUnitsMap(paramsBPC, dataTablesBPC as any, unitSystem)
    expect(result).toMatchSnapshot()
  })

  it('returns the unit map correctly with multiple units in datatables', () => {
    const unitSystem = 'Metric'

    const result = getParamUnitsMap(paramsCharpy, dataTablesCharpy as any, unitSystem)
    expect(result).toMatchSnapshot()
  })

  it('returns an empty unit map when params and dataTables are empty', () => {
    const emptyParams: any[] = []
    const emptyDataTables: any[] = []
    const unitSystem = 'Metric'

    const expected: UnitMap = {}

    const result = getParamUnitsMap(emptyParams, emptyDataTables, unitSystem)
    expect(result).toEqual(expected)
  })
})
