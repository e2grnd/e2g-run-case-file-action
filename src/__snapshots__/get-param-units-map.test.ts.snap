// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParamUnitsMap returns the unit map correctly 1`] = `
{
  "AXIAL_AREA_PARAMS_DOWNSTREAM": [
    "mm^2",
  ],
  "AXIAL_AREA_PARAMS_UPSTREAM": [
    "mm^2",
  ],
  "AXIAL_LENGTH_PARAMS_DOWNSTREAM": [
    "mm",
  ],
  "AXIAL_LENGTH_PARAMS_UPSTREAM": [
    "mm",
  ],
  "AXIAL_PERCENT": [
    "%",
  ],
  "AXIAL_POSITION": [
    "mm",
  ],
  "AXIAL_WALL_DISP": [
    "mm",
  ],
  "AllowableEffS": "kPa",
  "AllowableTmax": "kPa",
  "AllowableTmin": "kPa",
  "AxleLoad": "kN",
  "DSHr": "kPa",
  "DSLr": "kPa",
  "GRID_DATA": [
    "mm",
  ],
  "InBoredDiameter": "mm",
  "InDistanceOfGirthWeldFromTrackCenter": "m",
  "InFatigueResistanceOfGirthWeld": "kPa",
  "InInstallationTemp": "C",
  "InLengthOfRailroadTrack": "m",
  "InMaximumOperatingTemp": "C",
  "InMinimumOperatingTemp": "C",
  "InOperatingPressure": "kPa",
  "InOperatingWeightPerVehiclePerTrack": "kN",
  "InVehicleContactArea": "m^2",
  "InVerticalDepthOfSoilCover": "m",
  "InWallThickness": "mm",
  "InWidthOfRailroadTrack": "m",
  "MAX_PRESSURE": [
    "MPa",
  ],
  "MIN_PRESSURE": [
    "MPa",
  ],
  "PRESSURE_MEAN_COL": [
    "MPa",
  ],
  "PRESSURE_RANGE_COL": [
    "MPa",
  ],
  "PRES_HIST_1": [
    "MPa",
  ],
  "PRES_HIST_2": [
    "MPa",
  ],
  "PRES_TIME_1": [
    "yrs",
  ],
  "PRES_TIME_2": [
    "yrs",
  ],
  "S1": "kPa",
  "S2max": "kPa",
  "S2min": "kPa",
  "S3": "kPa",
  "SHe": "kPa",
  "SHi": "kPa",
  "SHiB": "kPa",
  "SHiBAllowableTmax": "kPa",
  "SHiBAllowableTmin": "kPa",
  "STRESS_RANGE_COL": [
    "MPa",
  ],
  "Seffmax": "kPa",
  "Seffmin": "kPa",
  "TRANSVERSE_AREA_PARAMS_CCW": [
    "mm^2",
  ],
  "TRANSVERSE_AREA_PARAMS_CW": [
    "mm^2",
  ],
  "TRANSVERSE_LENGTH_PARAMS_CCW": [
    "mm",
  ],
  "TRANSVERSE_LENGTH_PARAMS_CW": [
    "mm",
  ],
  "TRANSVERSE_PERCENT": [
    "%",
  ],
  "TRANSVERSE_POSITION": [
    "mm",
  ],
  "TRANSVERSE_WALL_DISP": [
    "mm",
  ],
  "W": "kPa",
  "YEARLY_NUMBER_OF_CYCLES": [
    "DimensionlessMetric",
  ],
  "outerDiameter": "mm",
}
`;

exports[`getParamUnitsMap returns the unit map correctly with multiple units in datatables 1`] = `
{
  "A": "J",
  "B": "J",
  "C": "C",
  "CHARPY_DATATABLE": [
    "C",
    "J",
    "mm",
    "DimensionlessMetric",
  ],
  "CTTCON": "C",
  "CVD1": [
    "C",
  ],
  "CVD2": [
    "J",
  ],
  "CVD3": [
    "mm",
  ],
  "CVNCON": "N-m",
  "CVNLOW": "N-m",
  "CVNMAX": "J",
  "CVNMIN": "J",
  "CVNTTEM": "C",
  "CVNUP": "N-m",
  "D": "C",
  "MASTER_T0": "C",
  "SSE": "J^2",
  "SSM": "J^2",
  "TT_15": "C",
  "TT_20": "C",
  "TT_20J": "C",
  "TT_28J": "C",
  "TT_40": "C",
  "TT_54J": "C",
}
`;
